allennlp/data/token_indexers/wordpiece_indexer.py

```diff
@@ -245,4 +256,25 @@ def __init__(self,
                          do_lowercase=do_lowercase,
                          never_lowercase=never_lowercase,
                          start_tokens=["[CLS]"],
-                         end_tokens=["[SEP]"])
+                         end_tokens=["[SEP]"],
+                         separator_token="[SEP]")
+
+
+def _get_token_type_ids(wordpiece_ids: List[int],
+                        separator_ids: List[int]) -> List[int]:
+    length_wordpiece = len(wordpiece_ids)
+    length_end_pieces = len(separator_ids)
+    token_type_ids = [0] * length_wordpiece
+    type_id = 0
+    cursor = 0
+    while cursor < len(wordpiece_ids):
+        if separator_ids[0] != wordpiece_ids[cursor]:
+            cursor += 1
+        else:
+            if separator_ids == wordpiece_ids[cursor:cursor+length_end_pieces]:
+                type_id += 1
+                cursor += length_end_pieces
+                token_type_ids[cursor:] = [type_id] * (length_wordpiece - cursor)
+            else:
+                cursor += length_end_pieces
```

Comment:
in practice, the separator is going to have length 1. I would probably try to avoid taking slices here, as it's inefficient. I think something like this would be more readable (and probably faster):

```
token_type_ids = []
cursor = 0
type_id = 0

if all(wordpiece_ids[cursor + i] == separator_id for i, separator_id in enumerate(separator_ids)):
    type_id += 1
    cursor += len(separator_ids)
    token_type_ids.extend(type_id for _ in separator_ids)
else:
    cursor += 1
    token_type_ids.append(type_id)
```

I think that does the same thing and is slightly more explicit and doesn't create unnecessary lists along the way.